### ISSUE TITLE
wish list: mcmc

### ISSUE THREAD

**Issue by mebrooks** [kaskr/adcomp#173]
Thanks @colemonnahan for the mcmc sampler development. This will be very useful. I have a couple of suggestions or wish list items for this functionality. 
~~1. It would be great to get samples of reported variables. Maybe a flag in the mcmc function could tell it to return values from `obj$report()`.~~
2. An example of how to get samples in parallel using `library(parallel)`. I will work on this one. I think it may be as simple as running different chains on different cores.

**Comment by mebrooks** [comment]
I just thought of another one...
3. A progress bar to show how close the sampler is is to completing the job.

**Comment by colemonnahan** [comment]
Hi @mebrooks, 

I think the biggest improvement is getting box constraints working. These need to be passed as arguments like the optimizers. The math gets a little tricky because of how I've dealt with the mass matrix: it's used in a linear transformation which, as far as I can tell, complicates the Jacobian calculations of the bounding function. I would love some extra brain power to help with that at some point too.

As for your questions:
1. As far as I know this requires passing the MCMC iterations through `obj$report()`, essentially looping through each iteration. I figured the user could do that after-the-fact, in case any post hoc thinning/burnin was done. However, it would be trivial to add that into the top level `mcmc` function an argument and would be a good idea. The only issue I can see is in the shape of the report data... a user might return a scalar, vector and matrix for each iteration. Automatically and robustly formatting that into a matrix would take some thought.

2. @kaskr and I went back and forth on this a bit. He prefers not to have parallel functionality built in b/c of platform differences and the user can add parallel into the model itself. It might be worth showing an example of how to run parallel chains using existing R packages, it's a bit different between platforms (hardest on Windows) but easily done. You might consider discussing it with him before putting any work into it.

3. A colleague here in Seattle did a pull request on my cloned repo that does this. I haven't taken the time to merge it in yet but it's also a good idea. My idea was to copy the `rstan` console output format.

BTW I'm following the git workflow from here: https://github.com/kaskr/adcomp/wiki/Development#recommended-workflow-fork-and-pull-request

**Comment by mebrooks** [comment]
I think you're right about item 1 above. I think it's best to let the user deal with it after thinning. So I edited it to be crossed out.

**Comment by colemonnahan** [comment]
This came up in the TMB workshop in Seattle too. I think all we need to do is give an example of how to do this, and state it should be done post hoc after verifying the chain has converged, etc.

There's an example MCMC file that would be the best place to put it:

https://github.com/kaskr/adcomp/blob/master/TMB/inst/examples/mcmc_examples.R

**Comment by iagomosqueira** [comment]
The mcmc samplers look really good, thanks @colemonnahan for this.

I am hoping to use TMB with mcmc on an existing package currently dependant on [coda](https://cran.r-project.org/web/packages/coda/index.html), which uses the `mcmc` S3 class. So coda and TMB conflict right now over the `mcmc` function name, which in coda creates an object of class `mcmc` from a data.frame. Which one works depends purely on the order the packages get loaded.

```
> library(TMB)
> library(coda)

Attaching package: ‘coda’

The following object is masked from ‘package:TMB’:

    mcmc
```
> 




Short of TMB importing coda and being fully compatible with it in outputs, maybe the mcmc function name could be changed to avoid this? This is becoming a common issue in R, and it is pretty much impossible to avoid conflicts across all packages, but coda is being around for a long time and it is widely used.

**Comment by kaskr** [comment]
Shouldn't be a problem as you can use ```TMB::mcmc```.

**Comment by iagomosqueira** [comment]
True, possible although tedious on an interactive session. A bit more problematic on existing code. In case you thought it was still worth considering.

**Comment by John-R-Wallace** [comment]
An alias can be created, e.g:

      Mcmc <- function(...) TMB::mcmc(...)

to be used in an interactive session or added to the top of existing 
code, or perhaps added to TMB.

Mcmc() works on the simple example as does copying the function:

Mcmc <- TMB::mcmc


FYI,

If the environment used is important then creating an alias may require 
additional information, e.g.:

     Ls <- function(name = 1, ...) base::ls(name = name, ...) # Works 
like base::ls() on the command line

     Ls <- function(...) base::ls(...) # Gives a list of the function's 
environment, not the global environment


-John

**Comment by colemonnahan** [comment]
I'm in favor or renaming the sampling function to something like `run_mcmc` or `sample_mcmc`. Almost all users will have `coda` loaded so they'll consistently use `TMB::mcmc` which just doesn't seem like a clean solution. I like the explicit namespace to avoid loading a package or an occasional conflict, but it seems the wrong tool in this case.

I'll be using TMB in a class next quarter so I'll be getting to some of these wish list items. Feel free to drop some more on here.

**Comment by kaskr** [comment]
Having had a look at existing mcmc R packages I now tend to agree with
@colemonnahan and @iagomosqueira that mcmc is *not* the right name
for the function that carries out the sampling. Given that it is clearly
stated that mcmc is in beta I'm OK with breaking backwards
compatibility, and simply renaming all the ```mcmc*``` functions to
```run_mcmc*``` (or perhaps ```tmb_mcmc*```?).
Any final preferences for the name ?

**Comment by colemonnahan** [comment]
I'd be fine with `run_mcmc`. 

While we're at it, it might be worth considering taking a close look at the `LaplacesDemon` R package which does Bayesian inference. It has a ton of algorithms, diagnostics, and other tools. The main issue is that you program your model in R so it's **slow**. However, if we could co-opt it to work with compiled TMB objects then it would make an extremely powerful framework with little work on the TMB end. 

It's no longer on CRAN but there's a repo: https://github.com/ecbrown/LaplacesDemon

In a sense the two packages are made for each other. FYI: LD has a simplified version of the NUTS and HMC algorithms (no mass matrix) so ours are currently better. But I'm guessing most algorithms we could plug right in. The Stan devs would probably argue they're all inferior to NUTS but who knows.

**Comment by iagomosqueira** [comment]
run_mcmc sounds good to me too.

**Comment by colemonnahan** [comment]
I'll be finally working on these issues. If anyone has any other requests now would be a good time to post them. @kaskr @mebrooks

**Comment by colemonnahan** [comment]
I added mass matrix adaptation, bounds, and parallel execution to my package https://github.com/colemonnahan/adnuts . @mebrooks did you ever try working with parallel? I used snowfall which works great for ADMB but with TMB there is no console output. Maybe there is a better way to do this, with another package? [See code here](https://github.com/colemonnahan/adnuts/blob/master/R/sample_tmb.R#L116). I'm not familiar with the differences and don't really want to spend time learning them. For now it seems to work well otherwise.

Can someone test it? Use [this file](https://github.com/colemonnahan/adnuts/blob/master/inst/mcmc_examples.R) as an example:

**Comment by mebrooks** [comment]
I use`parLapply` from the `parallel` package pretty often. `parallel` is nice because it works on all the OSs. Is there something specific I should try with it?

**Comment by mebrooks** [comment]
I see the parallel part of the example now. I was able to run it ok (on a Mac) and Andrea ran it in Windows. We both get this warning message:
```
Warning message:
In sample_tmb(obj = obj, iter = iter, chains = 3, seeds = 1:3, init = init,  :
  Note: Console output routed to mcmc_progress.txt when using parallel execution
```

[Here's](https://github.com/fishfollower/SAM/blob/ed5fd70c22efe4a47e0e2c54de447ae54a8c5b3e/stockassessment/R/retro.R#L93) an example of a function that uses `parLapply`. Does `snowfall` have some undesirable qualities and need to be replaced?

**Comment by colemonnahan** [comment]
I don't know the differences among parallel packages. The only reason I don't like snowfall is the output doesn't show up on the console and there doesn't appear to be a way to  get this easily. I just picked one and got it working. Perhaps later I can try parLapply. If someone else wants to try it I'd be happy to merge it in.

It's low priority for me at the moment since it works on Mac and Windows.

**Comment by kaskr** [comment]
A [new paper](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0197954) is out introducing the CRAN package [tmbstan](https://cran.r-project.org/web/packages/tmbstan/index.html) which can do MCMC for TMB model objects.
Checkout a simple example [here](https://github.com/kaskr/tmbstan#examples).

**Comment by mebrooks** [comment]
I skimmed the paper and this is very cool! So I guess this issue can be closed, right? 

On the wildflower example in the paper...it's a nice example showing how to see when the Laplace is a bad approximation and your results agree with the biology. We suspected that the Laplace would be a bad approximation of the random effect of year because the response (reproduction) is synchronized in all the sites and bimodal in 2 sites (Crone & Lesica 2004).

**Comment by colemonnahan** [comment]
@mebrooks That's a really fascinating insight regarding the example. I haven't thought too deeply about  the model/data but it's great to have a biological interpretation for why this might occur. 

Yes you can close this issue (I don't have permissions)
